lms/services/vitalsource/client.py

```diff
@@ -52,9 +60,19 @@ def book_toc(self, book_id: str):
 
             raise
 
-        return BookTOCSchema(response).parse()
+        schema = BookTOCSchema(response)
+        schema.context["book_id"] = book_id
+        return schema.parse()
+
+    @staticmethod
+    def parse_document_url(document_url):
+        return DOCUMENT_URL_REGEX.search(document_url).groupdict()
+
+    @staticmethod
+    def generate_document_url(book_id, cfi):
+        return f"vitalsource://book/bookID/{book_id}/cfi/{cfi}"
```

Comment:
The CFI may contain characters like `;` and `[`. For example: `vitalsource://book/bookID/9781400847402/cfi//6/28[;vnd.vst.idref=ch5]`.

I wonder if we should be URI-encoding the whole CFI. I'm just reading up on the relevant specs to figure out what we "should" be doing.

Reply:
We are URL encoding it when it's going to be on an URL for example on SpeedGrader (well, we have to go around a Canvas bug there).

I think if we urlencoded it on the API we might double-encode it ourselves unless we do a bigger refactor.

We can revisit this of course if it turns out we should encode it.